tango/integrations/beaker/workspace.py

```diff
@@ -0,0 +1,278 @@
+            tmp_dir = Path(tmp_dir_name)
+
+            for step in all_steps:
+                if step.name is None:
+                    continue
+
+                step_info = self._get_or_set_step_info(step)
+                steps[step.name] = step_info
+                run_data[step.name] = step.unique_id
+
+            with open(tmp_dir / Constants.RUN_DATA_FNAME, "w") as f:
+                json.dump(run_data, f)
+
+            self.beaker.dataset.sync(run_dataset, tmp_dir / Constants.RUN_DATA_FNAME, quiet=True)
+
+        # Commit the run dataset. It won't need to change again.
+        self.beaker.dataset.commit(run_dataset)
+
+        return Run(name=cast(str, name), steps=steps, start_date=run_dataset.created)
+
+    def registered_runs(self) -> Dict[str, Run]:
+        runs: Dict[str, Run] = {}
+        # TODO: do these requests concurrently
+        for dataset in self.beaker.workspace.datasets(uncommitted=False):
+            if dataset.name is None:
+                continue
+            if dataset.name.startswith(Constants.RUN_DATASET_PREFIX):
+                run = self._get_run_from_dataset(dataset)
+                runs[run.name] = run
+        return runs
+
+    def registered_run(self, name: str) -> Run:
+        try:
+            dataset_for_run = self.beaker.dataset.get(run_dataset_name(name))
+            if dataset_for_run.workspace_ref.id != self.beaker.workspace.get().id:
+                raise DatasetNotFound
+        except DatasetNotFound:
+            raise KeyError(f"Run '{name}' not found in workspace")
+        return self._get_run_from_dataset(dataset_for_run)
+
+    def _get_run_from_dataset(self, dataset: Dataset) -> Run:
+        assert dataset.name is not None
+        run_name = dataset.name[len(Constants.RUN_DATASET_PREFIX) :]
+        steps: Dict[str, StepInfo] = {}
+        steps_info_bytes = b"".join(
+            self.beaker.dataset.stream_file(dataset, Constants.RUN_DATA_FNAME, quiet=True)
+        )
+        steps_info = json.loads(steps_info_bytes)
+        # TODO: do these requests concurrently
+        for step_name, unique_id in steps_info.items():
+            steps[step_name] = self.step_info(unique_id)
+        return Run(name=run_name, start_date=dataset.created, steps=steps)
+
+    def _get_or_set_step_info(self, step: Step) -> StepInfo:
+        try:
+            return self.step_info(step)
+        except KeyError:
+            step_info = StepInfo.new_from_step(step)
+            self._update_step_info(step_info)
+            return step_info
+
+    def _update_step_info(self, step_info: StepInfo):
+        dataset_name = step_dataset_name(step_info)
+
+        step_info_dataset: Dataset
+        try:
+            step_info_dataset = self.beaker.dataset.create(dataset_name, commit=False)
+        except DatasetConflict:
+            step_info_dataset = self.beaker.dataset.get(dataset_name)
+
+        if step_info_dataset.committed:
+            return
```

Comment:
What does this mean? If it's committed, we silently don't update the step info?

Reply:
Yea.. I don't think we need that though. This shouldn't happen. 0eba6ab